### PR TITLE
Contribute the P4Tools platform and P4Testgen 

### PR DESCRIPTION
**Pull request by fruffy-bfn**
This is the initial commit to add P4Testgen ([Video](https://www.youtube.com/watch?v=yucl3azbh8Q), [Slides](https://opennetworking.org/wp-content/uploads/2022/05/Fabian-Ruffy-Final-Slide-Deck.pdf)) as a back end to P4C. This is rather large pull request, however all the parts included are necessary to generate proper tests using P4Testgen.

Description:

# P4Tools
P4Tools is a collection of tools that make testing P4 targets and programs a little easier. So far the platform supports the following tools and projects:

- [P4Testgen](https://github.com/fruffy-bfn/p4c/tree/p4tools/backends/p4tools/testgen): An input-output test case generator for P4.

### Building

P4Tools can be built using the following CMAKE configuration in the P4C repository.

```
mkdir build
cd build
cmake .. -DENABLE_GMP=OFF -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DENABLE_TOOLS=ON
make
```

### Dependencies
* [inja](https://github.com/pantor/inja) template engine for testcase generation.
* [gsl-lite](https://github.com/gsl-lite/gsl-lite) C++ Core Guidelines Support Library, used to
support and enforce C++ best practices.
* [z3](https://github.com/Z3Prover/z3) P4Tools requires a Z3 version that is at least 4.8.14. 

# P4Testgen - Generating Tests for P4 Targets

The P4Testgen tool uses symbolic execution to automatically generate input-output tests for a given P4 program. P4Testgen attempts to generate a set of tests of all reachable paths within the program. Each test consists of an input packet, the control-plane commands to populate match-action tables, and a predicate on the output packet. Each generated test also has a human-readable trace that indicates the path through the program being exercised by that test.

Currently, P4Testgen supports generating tests for the v1model architecture of the behavioral model. 

## Usage
After building, the main binary can be found in `build/p4check`.

To generate tests for a particular v1model program run
```
build/p4check testgen --target bmv2 --arch v1model --std p4-16 -Ibuild/p4include --test-backend STF --print-traces --seed 1000 --max-tests 10 --out-dir build/testgen/testgen-p4c-bmv2/bmv2_table_default_action.out backends/p4tools/testgen/targets/bmv2/test/p4-programs/bmv2_table_default_action.p4
```
`p4check` is an umbrella binary which delegates execution to all tools of the p4-tools repository.
`testgen` specifies that the testgen tool should be used. In the future, other tools may be supported, for example random program generation (smith, mutate) or translation validators.

Once P4Testgen has generated tests, the tests can be executed by either the P4Runtime or STF test back ends.

### Running ctest
Once compiled, run ctest in the build directory in the following way : `ctest -V -R testgen-` 

This will execute all the tests that are currently related to P4Testgen.

### CODE REVIEW

**Comment by fruffy-bfn** [comment]
@mbudiu-vmw This pull request is basically ready to go. I tried to add p4testgen in a non-invasive way which should not affect the rest of the compiler. We require an additional CI run but this run takes roughly the same amount as the other test runs. The CI run compiles p4tools/p4testgen and then generates 10 tests for each v1model bmv2 P4 program. 

To reduce the burden on CI I introduced #3485  and #3491 which will substantially reduce the number of CI runs. However, until they are ready it might take a little while.

Please let us know if you have any questions and concerns related to this pull request. Unfortunately, I do not think we can break it into smaller parts since they will negatively impact the remaining development workflow. However, we will consolidate and contribute parts that are duplicated and/or redundant over the next couple of weeks, once the tool is merged. 

I will also try to get other compiler developers to take a look over this PR, considering how large it is.

**Review by jnfoster — APPROVED** [review]
🎉🎉🎉
🚀🚀🚀

I confess I have only skimmed this mega PR -- I haven't reviewed the code with a fine-tooth comb. However, I've been a part of the team developing this code for over a year, so I know it well. And I just confirmed that I can build it from a fresh clone on Mac OS. So I think this is good to go!

Congrats @fruffy-bfn! What a great milestone!

File: backends/p4tools/cmake/FindZ3.cmake
Comment (rst0git) [inline backends/p4tools/cmake/FindZ3.cmake:13]: The following change would allow to build with `libz3.so` (e.g., when installed from a package)
```Suggestion
  NAMES z3
```

**Review by rst0git — APPROVED** [review]

File: CMakeLists.txt
Comment (mbudiu-vmw) [inline CMakeLists.txt:38]: "TOOLS" is a very generic name. Should this be called TEST_TOOLS?

File: backends/p4tools/README.md
Comment (mbudiu-vmw) [inline backends/p4tools/README.md:13]: can't we reuse the existing cpplint tool?

File: backends/p4tools/README.md
Comment (mbudiu-vmw) [inline backends/p4tools/README.md:58]: that is unfortunate and too late to change now.

File: backends/p4tools/cmake/common.cmake
Comment (mbudiu-vmw) [inline backends/p4tools/cmake/common.cmake:1]: so is this P4tools or p4testgen? I prefer the latter.

File: backends/p4tools/common/compiler/configuration.h
Comment (mbudiu-vmw) [inline backends/p4tools/common/compiler/configuration.h:11]: it is a limit, but larger

File: backends/p4tools/testgen/testgen.h
Comment (mbudiu-vmw) [inline backends/p4tools/testgen/testgen.h:15]: not a particularly useful comment.

File: backends/p4tools/common/compiler/convert_varbits.h
Comment (mbudiu-vmw) [inline backends/p4tools/common/compiler/convert_varbits.h:11]: Nothing precludes the runtime implementation of varbits from doing this. In fact, it does in bmv2.

**Review by mbudiu-vmw — COMMENTED** [review]
At the rate I am reading code I expect this will take a couple of months to review at least.
I hope that there will be an attempt to fix the issues found. Otherwise I am not sure that the review makes sense. Here are some comments on a few files I read.

I really think that it would have been better to do this incrementally.

File: backends/p4tools/README.md
Comment (fruffy-bfn) [reply backends/p4tools/README.md:13]: Yes, I would do this in a separate PR. We can move up a lot of the tools here, they are generally useful.

File: backends/p4tools/README.md
Comment (fruffy-bfn) [reply backends/p4tools/README.md:58]: Actually, I do not think the differences are too big. However, we have more automated tools to check for style, autoformatting, etc. For example, we run https://github.com/fruffy-bfn/p4c/blob/p4tools/backends/p4tools/.clang-format for the PRs to automatically check code formatting. We can do this for the entire compiler. It would just require one big commit that autoformats the code.

Similarly, we run [IWYU](https://github.com/include-what-you-use/include-what-you-use) to clean up all the includes.

File: backends/p4tools/common/compiler/convert_varbits.h
Comment (fruffy-bfn) [reply backends/p4tools/common/compiler/convert_varbits.h:11]: We need a varbit type that retains the original max-width of the varbit and also contains information about the current assigned width. I do not think this is possible with the current varbit type? How does BMv2 do it? It would simplify the code a lot.

File: backends/p4tools/cmake/common.cmake
Comment (fruffy-bfn) [reply backends/p4tools/cmake/common.cmake]: It is P4Tools. P4Testgen is one tool contained within P4Tools. Currently, the only one but there will be more.

**Comment by fruffy-bfn** [comment]
> At the rate I am reading code I expect this will take a couple of months to review at least. I hope that there will be an attempt to fix the issues found. Otherwise I am not sure that the review makes sense. Here are some comments on a few files I read.
> 
> I really think that it would have been better to do this incrementally.

Yes, it is unfortunate, but it is hard to break this down into smaller parts. The code will make little sense and we are blocked on further development until all parts are merged in. We could segment it into the `common`, `testgen`, and `testgen/target` repositories?

File: backends/p4tools/common/compiler/convert_varbits.h
Comment (mbudiu-vmw) [reply backends/p4tools/common/compiler/convert_varbits.h:11]: Quoting from https://p4.org/wp-content/uploads/2022/07/P4-16-spec.html#sec-varbit-string

> Each occurrence of the type varbit has a statically-declared maximum width, as well as a dynamic width, which must not exceed the static bound. Prior to initialization a variable-size bit-string has an unknown dynamic width.
> 
> Variable-length bit-strings support a limited set of operations:
> 
>     Assignment to another variable-sized bit-string. The target of the assignment must have the same static width as the source. When executed, the assignment sets the dynamic width of the target to the dynamic width of the source.
>     Comparison for equality or inequality with another varbit field. Two varbit fields can be compared only if they have the same type. Two varbits are equal if they have the same dynamic width and all the bits up to the dynamic width are the same. 

Isn't this exactly what you need?

File: backends/p4tools/scripts/tools/cpplint.py
Comment (mbudiu-vmw) [inline backends/p4tools/scripts/tools/cpplint.py:1]: do we really need two copies of cpplint?
I can understand we need two configurations, but we should be able to reuse the tool.

File: backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp
Comment (mbudiu-vmw) [inline backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp:1]: what are these imp files? Do they need to be in the repository?

File: backends/p4tools/scripts/tools/run-clang-format
Comment (mbudiu-vmw) [inline backends/p4tools/scripts/tools/run-clang-format:4]: can we give a .sh suffix to this file to make it clear that it's a shell script?

File: backends/p4tools/testgen/targets/bmv2/register.h
Comment (mbudiu-vmw) [inline backends/p4tools/testgen/targets/bmv2/register.h:15]: In general I would recommend a little comment in each file describing its purpose.
For example, here based on the file name I thought this handles v1model Registers. 
It seems to be something quite different.

File: backends/p4tools/testgen/targets/bmv2/test/BMV2ProtobufXfail.cmake
Comment (mbudiu-vmw) [inline backends/p4tools/testgen/targets/bmv2/test/BMV2ProtobufXfail.cmake:116]: do they need it? if not, we can remove the include.

File: backends/p4tools/testgen/targets/bmv2/test/p4-programs/bmv2_action_run.p4
Comment (mbudiu-vmw) [inline backends/p4tools/testgen/targets/bmv2/test/p4-programs/bmv2_action_run.p4:1]: I bet it wouldn't have been too difficult to make all these test programs part of a separate PR.
But now I have started reviewing this one, let's not change this, or I will have to start from scratch.

**Review by mbudiu-vmw — COMMENTED** [review]
I read a few more files.
I don't plan to read the .p4 files, I assume that they are validated enough by the tools.

File: backends/p4tools/common/compiler/convert_varbits.h
Comment (fruffy) [reply backends/p4tools/common/compiler/convert_varbits.h:11]: Yes, afaik the `IR::Type_Varbits` does only have one size field. Even if we generated a new Type_Varbits, we would lose information about the original size and the types are not equivalent.

File: backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp
Comment (fruffy) [reply backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp:1]: They are mapping files that are used by IWYU. They are quite useful because they help with generating the right includes when using the tool. Without them, IWYU will not be usable.

File: backends/p4tools/testgen/targets/bmv2/test/BMV2ProtobufXfail.cmake
Comment (fruffy) [reply backends/p4tools/testgen/targets/bmv2/test/BMV2ProtobufXfail.cmake:116]: I was thinking that we patch those up once the tool is merged in. We can also move all the sample programs we have into the test folder.

File: backends/p4tools/common/compiler/convert_varbits.h
Comment (mbudiu-vmw) [reply backends/p4tools/common/compiler/convert_varbits.h:11]: The max size is compile-time information.
The compiler is not concerned with the dynamic size, that only exists at runtime.
I still fail to see why you need your encoding.

File: backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp
Comment (mbudiu-vmw) [reply backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp:1]: did you write them by hand?  if they are generated by a tool we should use the tool to generate the files.
or perhaps they come from a repository?

File: backends/p4tools/testgen/targets/bmv2/test/BMV2Xfail.cmake
Comment (fruffy) [reply backends/p4tools/testgen/targets/bmv2/test/BMV2Xfail.cmake:3]: `1. P4C Toolchain Issues ` describes issues we have found with either the compiler or the simple switch model. We have filed these. There are probably more, but we are still just running a limited amount of tests on the programs.

File: backends/p4tools/smith/smith.h
Comment (fruffy) [reply backends/p4tools/smith/smith.h:12]: I can remove this.

File: backends/p4tools/common/compiler/convert_varbits.h
Comment (fruffy) [reply backends/p4tools/common/compiler/convert_varbits.h:11]: Our interpreter assigns a dynamic size [here](https://github.com/p4lang/p4c/pull/3495/files#diff-ed409c5ff044fe5d8cafe027e7ef5007f433def7e39bd28974d9f3253c7be1aeR36) when we call an extract. At this point the varbit will have both information about the original maximum size and the dynamic size. We use the maximum size as means of comparison to avoid type mismatch issues.  
https://github.com/p4lang/p4c/pull/3495/files#diff-bfbb8698949a4875ee7cd51ba9a17c7a58f9a7a6a1f4542455a3b551f483ba30R18
(I will need to update that comment)

File: backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp
Comment (fruffy) [reply backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp:1]: https://github.com/include-what-you-use/include-what-you-use
They are handpicked from here. We could potentially link against this module and run the tool from there. 
`p4c.imp` is written by us.

File: backends/p4tools/testgen/targets/bmv2/test/p4-programs/bmv2_action_run.p4
Comment (fruffy) [reply backends/p4tools/testgen/targets/bmv2/test/p4-programs/bmv2_action_run.p4:1]: Yes, but then we would have not known whether the tool executes these program correctly. I will create a PR that contributes them to the testdata repository.

File: backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp
Comment (rst0git) [reply backends/p4tools/scripts/tools/iwyu_mappings/gcc.libc.imp:1]: @fruffy It might be useful to add a `Makefile` that would make it easier to update these files in the future:

```Makefile
all: gcc.libc.imp libcxx.imp stl.c.headers.imp stl.c.headers.imp boost-1.75-all.imp

gcc.libc.imp:
	curl -s https://raw.githubusercontent.com/include-what-you-use/include-what-you-use/master/gcc.libc.imp -o $@

libcxx.imp:
	curl -s https://raw.githubusercontent.com/include-what-you-use/include-what-you-use/master/libcxx.imp -o $@

stl.c.headers.imp:
	curl -s https://raw.githubusercontent.com/include-what-you-use/include-what-you-use/master/stl.c.headers.imp -o $@

boost-all.imp:
	curl -s https://raw.githubusercontent.com/include-what-you-use/include-what-you-use/master/boost-all.imp -o $@

boost-1.75-all.imp:
	curl -s https://raw.githubusercontent.com/include-what-you-use/include-what-you-use/master/boost-1.75-all.imp -o $@
```

It would be useful to explain how these files are used in `README.md`.

File: backends/p4tools/testgen/targets/bmv2/test/p4-programs/bmv2_action_run.p4
Comment (fruffy-bfn) [reply backends/p4tools/testgen/targets/bmv2/test/p4-programs/bmv2_action_run.p4:1]: On second thought, I am not entirely sure whether it makes sense to include them. They might add unnecessary CI time. They are not testing very unique cases.

File: backends/p4tools/testgen/targets/bmv2/register.h
Comment (fruffy-bfn) [reply backends/p4tools/testgen/targets/bmv2/register.h:16]: In general, we try to add comments to each function definition but sometimes things slip through. I am hoping that we are able to refactor this whole structure very soon to make it a bit simpler. It has grown over time and is starting to become confusing.

File: backends/p4tools/scripts/tools/iwyu_mappings/Makefile
Comment (rst0git) [inline backends/p4tools/scripts/tools/iwyu_mappings/Makefile:23]:
```Makefile
.PHONY: all clean refresh
```

File: backends/p4tools/testgen/test/lib/taint.cpp
Comment (mbudiu-vmw) [inline backends/p4tools/testgen/test/lib/taint.cpp:24]: since taint has a width, perhaps you should indicate it here. `taint<8>`?

File: backends/p4tools/testgen/backend/tf.h
Comment (mbudiu-vmw) [inline backends/p4tools/testgen/backend/tf.h:16]: TF = Test Framework?

Comment (mbudiu-vmw) [reply backends/p4tools/testgen/backend/tf.h:16]: This comment seems wrong to me. Isn't this a base class that should be overridden? If that's correct the constructor should be protected.

File: backends/p4tools/scripts/tools/README.md
Comment (mbudiu-vmw) [inline backends/p4tools/scripts/tools/README.md:6]: remove "checks"

File: backends/p4tools/scripts/hooks/pre-commit
Comment (mbudiu-vmw) [inline backends/p4tools/scripts/hooks/pre-commit:5]: maybe we should do this for the compiler too

File: backends/p4tools/common/options.cpp
Comment (mbudiu-vmw) [inline backends/p4tools/common/options.cpp:164]: can't this be done by inheriting another Options class?

File: backends/p4tools/testgen/targets/bmv2/target.cpp
Comment (mbudiu-vmw) [inline backends/p4tools/testgen/targets/bmv2/target.cpp:39]: I actually would very much like to automate this step -- see https://github.com/p4lang/p4c/pull/3281
This should make it much easier to write tools specific to a model.

Comment (mbudiu-vmw) [reply backends/p4tools/testgen/targets/bmv2/target.cpp:39]: Maybe if you have no other project going on we could team up?

File: backends/p4tools/testgen/lib/exceptions.h
Comment (mbudiu-vmw) [inline backends/p4tools/testgen/lib/exceptions.h:10]: I think this file should be called "unimplemented.h".

File: backends/p4tools/testgen/core/target.h
Comment (mbudiu-vmw) [inline backends/p4tools/testgen/core/target.h:32]: this could be automated: https://github.com/p4lang/p4c/pull/3281, I have left another comment about this.

File: backends/p4tools/common/core/solver.h
Comment (mbudiu-vmw) [inline backends/p4tools/common/core/solver.h:52]: was thinking to move this to a base class that everyone who wants to do "to" could inherit from. including in the compiler.

File: backends/p4tools/common/lib/coverage.cpp
Comment (mbudiu-vmw) [inline backends/p4tools/common/lib/coverage.cpp:13]: can't this be done for the Statement class?

**Review by mbudiu-vmw — COMMENTED** [review]
So far the code looks great, I went through about 160/250 files, but most of the work is still aheda of me. This is another collection of small comments.

File: backends/p4tools/common/compiler/convert_errors.h
Comment (fruffy-bfn) [reply backends/p4tools/common/compiler/convert_errors.h:20]: It is a modification of the convert_enums pass, there is no convert_errors pass. We were not able to consolidate those two passes, sadly. Maybe there is a good way? We can move up all these passes in the midend, since they are generally useful.

File: backends/p4tools/common/compiler/copy_headers.h
Comment (fruffy-bfn) [reply backends/p4tools/common/compiler/copy_headers.h:18]: It extends the copyStructures pass. I think it might be possible to merge these two.

File: backends/p4tools/common/lib/coverage.cpp
Comment (fruffy-bfn) [reply backends/p4tools/common/lib/coverage.cpp:13]: I believe this was done because not all statements are relevant. For example, switch or block statements are only wrappers for the assignment and method call statements contained within.

File: backends/p4tools/common/compiler/copy_headers.h
Comment (fruffy-bfn) [reply backends/p4tools/common/compiler/copy_headers.h:18]: I will add an issue for both of the passes, so we can make sure to add them to the midend folder later and potentially merge or fix them.

File: backends/p4tools/testgen/targets/bmv2/test/STFTestTemplate.cmake
Comment (fruffy-bfn) [reply backends/p4tools/testgen/targets/bmv2/test/STFTestTemplate.cmake:1]: This script automatically runs STF tests on the tests generated by P4Testgen. The BMv2 back end also supports the generating of PTF and Protobuf tests, but we do not have a test runner for these. I think these CMake files can also 
be cleaned up.

File: backends/p4tools/common/options.cpp
Comment (fruffy-bfn) [reply backends/p4tools/common/options.cpp:164]: I am actually not sure why this options class was written this way. Let me see.

File: backends/p4tools/testgen/lib/exceptions.h
Comment (fruffy-bfn) [reply backends/p4tools/testgen/lib/exceptions.h:10]: There is a plan to add finer-grained exceptions. For example, exceptions that indicate that the tool has encountered a path that is not valid, exceptions for unimplemented features, or exceptions that are fatal. We are still using COMPILER_BUG, but those messages are confusing.

File: backends/p4tools/common/core/solver.h
Comment (fruffy-bfn) [reply backends/p4tools/common/core/solver.h:52]: Yes, we can do this. But there are some includes such as model and formulae.h that make this class a little opinionated.

File: backends/p4tools/testgen/targets/bmv2/target.cpp
Comment (fruffy) [reply backends/p4tools/testgen/targets/bmv2/target.cpp:39]: Yes, something like this is sorely needed for testgen because we have to map the structure of the package to the nodes of the parsed P4 program.. But I am not entirely sure how we would consume this model yet. 

I am happy to contribute to this, but I unfortunately have little time.

File: backends/p4tools/testgen/targets/bmv2/constants.cpp
Comment (fruffy) [reply backends/p4tools/testgen/targets/bmv2/constants.cpp:1]: Some constants were moved up and now the file is empty. It is probably useful to keep it when you have to deal with constants that can not be defined in a header-only file.

**Comment by fruffy-bfn** [comment]
@VolodymyrPeschanenkoIntel  issue281 is failing because of a bug in loopsUnroll. The pass adds a OutofBounds transition although the parser simply contains an infinite loop.

File: backends/p4tools/common/options.cpp
Comment (fruffy-bfn) [reply backends/p4tools/common/options.cpp:164]: I tried simplifying this but unfortunately this requires a major rewrite. These options are written this way to make it possible to have every target extension define its own options without needing to produce a different binary. 
`p4check testgen`, `p4check mutate`,  `p4check verify` can all have their own options.

**Pull request merged**
